### PR TITLE
🚀 Prepare for the second version of this awesome library

### LINKED ISSUE #69: Why does filter preserve keys?
What was the reason behind filter preserving the original array's keys? Is there a benefit to it? I'm most often filtering arrays where the key is an integer, so I'd expected filter to act more like it does in other languages creating a new array with values accessible by index.

I get that updating filter would lead to breaking changes, and there are probably times where the current implementation comes in handy, but it'd be useful to to have a function that filters an array and returns a new array that starts at zero.

If the maintainers feel like it would be useful and extend what phunctional is trying to do, I could work on an implementation.

### PR DESCRIPTION
**Pull request by rgomezcasas**
## Breaking changes
* [x] Php >= 7.2
* [x] ~~`filter`, `flat_map` and `map` no longer sends the `key` as a second argument. Use `filter_indexed`, `flat_map_indexed` and `map_indexed` for the old behaviour.~~ `filter`, `map` and `flat_map` now supports the opcionallity of the key as the second argument. (fixes #69, closes #70)

## Why
This is the cumulation of a lot of feedback received this last year.

### CODE REVIEW

File: tests/FilterTest.php [hunk 1]
```diff
@@ -1,40 +1,26 @@
...
-use PHPUnit_Framework_TestCase;
+use PHPUnit\Framework\TestCase;
 use function Lambdish\Phunctional\filter;
 
-class FilterTest extends PHPUnit_Framework_TestCase
+final class FilterTest extends TestCase
 {
     /** @test */
-    public function it_should_filter_a_collection_keeping_the_indexes()
+    public function it_should_filter_a_collection_keeping_the_indexes(): void
```
Comment (jorgeavilacardenosa) [inline tests/FilterTest.php]: Can we test the asserting with an array and also with an iterator?

$actual = new ArrayIterator(['one' => 1, 'two' => 2, 'three' => 3, 'four' => 4, 'five' => 5]);
$pred   = $this->evenByValueFilterer();

$this->assertSame(['two' => 2, 'four' => 4], filter($pred, $actual));

File: src/instance_of.php [hunk 2]
```diff
@@ -1,19 +1,19 @@
...
 
 /**
  * Returns a checker that validated if an `$element` is an instance of a `$className`
  *
  * @param string $className class name to compare with
- *
- * @return Closure
  */
-function instance_of($className)
+function instance_of($className): callable
```
Comment (jorgeavilacardenosa) [inline src/instance_of.php]: function instance_of($className): callable

File: tests/MapTest.php [hunk 3]
```diff
@@ -17,34 +19,18 @@ public function it_should_return_the_values_of_a_collection_after_apply_a_functi
     }
 
     /** @test */
-    public function it_should_keep_the_keys()
+    public function it_should_keep_the_keys(): void
     {
         $actual   = ['one' => 1, 'two' => 2, 'three' => 3, 'four' => 4, 'five' => 5];
```
Comment (jorgeavilacardenosa) [inline tests/MapTest.php]: the same as before, we are not testing the collection implementing Traversable.

File: tests/MemoizeTest.php [hunk 4]
```diff
@@ -30,27 +32,12 @@ public function it_should_reset_the_cache_if_function_to_be_executed_is_null()
      * @test
      * @dataProvider fibonacciValues
      */
-    public function it_should_call_fibonacci($number, $fibonacci)
-    {
-        $this->assertSame($fibonacci, apply($this->functionFibonacciMemoized(), [$number]));
-    }
-
-    private function functionWithRandReturn()
+    public function it_should_call_fibonacci($number, $fibonacci): void
```
Comment (jorgeavilacardenosa) [inline tests/MemoizeTest.php]: maybe the testname is not the best :( 

and yeah, it's mine! xD

File: composer.json [hunk 5]
```diff
@@ -26,10 +26,11 @@
   ],
   "minimum-stability": "beta",
   "require": {
-    "php": ">=5.6.0"
+    "php": ">=7.2"
```
[the remark this answers is not in the document]
Comment (rgomezcasas) [reply composer.json]: I updated to 7.2, but it's already [almost deprecated](https://www.php.net/supported-versions.php)

Usually people that don't upgrade the php server also doesn't upgrade its dependencies xD

**Pull request merged**